src/core_landice/mode_forward/mpas_li_calving.F

```diff
@@ -1249,6 +1256,794 @@ subroutine eigencalving(domain, err)
+
+         !call mpas_pool_get_field(scratchPool, 'epsmaxField', epsmaxField )
+         !call mpas_allocate_scratch_field(epsmaxField, .true.)
+         !epsmax => epsmaxField % array
+
+         !call mpas_pool_get_field(scratchPool, 'nstarField', nstarField )
+         !call mpas_allocate_scratch_field(nstarField, .true.)
+         !nstar => nstarField % array
+
+         !call mpas_pool_get_field(scratchPool, 'szeroField', szeroField )
+         !call mpas_allocate_scratch_field(szeroField, .true.)
+         !szero => szeroField % array
+
+         !call mpas_pool_get_field(scratchPool, 'sourceField', sourceField )
+         !call mpas_allocate_scratch_field(sourceField, .true.)
+         !source => sourceField % array
+
+         !call mpas_pool_get_field(scratchPool, 'ddamagedtField', ddamagedtField )
+         !call mpas_allocate_scratch_field(ddamagedtField, .true.)
+         !ddamagedt => ddamagedtField % array
+
+         call mpas_pool_get_field(scratchPool, 'forceField', forceField )
+         call mpas_allocate_scratch_field(forceField, .true.)
+         force => forceField % array
+
+         call mpas_pool_get_field(scratchPool, 'meanTField', meanTField )
+         call mpas_allocate_scratch_field(meanTField, .true.)
+         meanT => meanTField % array
+
+         call mpas_pool_get_config(liConfigs, 'config_ice_density', config_ice_density)
+         call mpas_pool_get_config(liConfigs, 'config_ocean_density', config_ocean_density)
+         rhoi = config_ice_density
+         rhoo = config_ocean_density
+
+         meanT(:) = sum(temperature(:,:), dim=1)/nVertLevels
+
+         call li_compute_gradient_2d(meshPool, uReconstructX(1,:), exx, exy, err_tmp)
+         err = ior(err, err_tmp)
+
+         call li_compute_gradient_2d(meshPool, uReconstructY(1,:), eyx, eyy, err_tmp)
+         err = ior(err, err_tmp)
+
+         call li_calculate_flowParamA(meshPool, temperature, thickness, flowParamA, err_tmp)
+         err = ior(err, err_tmp)
+
+
+         meanFlowParamA(:) = sum(flowParamA(:,:), dim=1)/nVertLevels
+         ! calculate the depth-averaged flow parameter A
+
+         call mpas_log_write("meanFlowParamA: Min=$r, Max=$r", &
+                    realArgs=(/minval(meanFlowParamA), maxval(meanFlowParamA)/))
+         if (config_damage_use_constant_A) then
+             where (thickness == 0.0_RKIND)
+                 visc = 0.0_RKIND
+             elsewhere
+                 visc = 0.5*stiffnessFactor*config_default_flowParamA**(-1.0/3.0)*(sqrt(exx**2+eyy**2 + exx*eyy + 0.25*(exy+eyx)**2))**((1.0-3.0)/3.0)
+             endwhere
+         else
+             where (thickness == 0.0_RKIND)
+                 visc = 0.0_RKIND
+             elsewhere
+                 visc = 0.5*stiffnessFactor*meanFlowParamA**(-1.0/3.0)*(sqrt(exx**2+eyy**2 + exx*eyy + 0.25*(exy+eyx)**2))**((1.0-3.0)/3.0)
+             endwhere
+         end if
+         ! recover effective viscosity from vel. and temp. fields
```

Comment:
It might make more sense to move all this viscosity and stress info to its own routine that is independent of the damage code. @stephenprice , what do you think?  It might make more sense in the velocity module.  Eventually, we might get this info from Albany and not need to calculate it ourselves.

Reply:
Yeah, eventually. It is better to just call those from Albany

Reply:
I think that is fine. I can imagine any number of other uses for this information external to damage / calving calculations so it makes sense to me to have it available in its own module.

Reply:
I think we should move the stress and viscosity calculations to subroutine `calculate_strain_rates` that already exists in `mpas_li_velocity.F`.  We could rename that routine to `calculate_strain_stress` or something.  I know that @trhille has a similar stress calculation somewhere in his von Mises calving code, so having one place where all this gets defined in a consistent way would be good.  Then when we finally get around to having these fields passed from Albany, we can replace this with that.

Reply:
Now use calculate_strain_rates subroutine to calculate exx, exy, eyx and eyy

Reply:
I might want to leave this (adding viscosity calculation in mpas_li_velocity), as here I need to decide if we are using a constant flow parameter or not, I am not quite confident how to put that in mpas_li_velocity.F. Actually we don't have to use the constant A. We can just use the A value given by the model, but we have to make sure we are getting the correct A from Albany.

Reply:
@tongzhangice , what do you use the config_damage_use_constant_A for in your simulations?  Are you running a configuration where you have A constant for damage, but not for the velocity solver?

Reply:
No, I don't have such experiments (variable A for vel. solver but constant for damage calculation). I can't remember exactly why I add this option. Maybe for some reasons I didn't get the correct A value from the li_calculate_flowParamA subroutine and I applied a constant A directly.

Reply:
In that case, I think we should delete this option and if there is a problem in the li_calculate_flowParamA routine, we can fix it there.

Reply:
the constant A option is deleted

Reply:
In the long term, we still might want to move these viscosity calculations to a different module, but given the issues with how that all works right now, we can leave what you have for now.